front/lib/credits/purchase.ts

```
@@ -0,0 +1,267 @@
+import { Err, Ok } from "@app/types";
+
+/**
+ * Handles credit purchase activation from a Stripe invoice for Pro subscriptions.
+ * (Enterprise subscriptions activate credits optimistically, so we skip them here.)
+ *
+ * @param invoice - The Stripe invoice that was paid
+ * @param subscription - The subscription record with workspace included
+ * @returns Result indicating success or failure
+ */
+export async function maybeStartCreditFromProOneOffInvoice({
+  invoice,
+  subscription,
+}: {
+  invoice: Stripe.Invoice;
+  subscription: Subscription & { workspace: WorkspaceModel };
+}): Promise<Result<undefined, Error>> {
+  // Check if this is a credit purchase invoice
+  if (!isCreditPurchaseInvoice(invoice)) {
+    // Not a credit purchase invoice, nothing to do
+    return new Ok(undefined);
+  }
+
+  if (typeof invoice.subscription !== "string") {
+    return new Err(
+      new Error(
+        "Credit purchase invoice does not have a valid subscription ID."
+      )
+    );
+  }
+
+  const workspace = subscription.workspace;
+  const creditAmountCents = getCreditAmountFromInvoice(invoice);
+
+  if (creditAmountCents === null) {
+    logger.error(
+      {
+        workspaceId: workspace.sId,
+        invoiceId: invoice.id,
+        creditAmountCents: invoice.metadata?.credit_amount_cents,
+      },
+      "[Credit Purchase] Invalid credit amount in invoice metadata"
+    );
+    return new Err(new Error("Invalid credit amount in invoice metadata"));
+  }
+
+  const stripeSubscription = await getStripeSubscription(invoice.subscription);
+  if (!stripeSubscription) {
+    logger.error(
+      {
+        workspaceId: workspace.sId,
+        invoiceId: invoice.id,
+      },
+      "[Credit Purchase] Subscription not found"
+    );
+    return new Err(new Error("Subscription not found"));
+  }
+  // Check if this is an Enterprise subscription
+  const isEnterprise = isEnterpriseSubscription(stripeSubscription);
+
+  if (isEnterprise) {
+    logger.info(
```

IMO we should throw an error here, this function should never be called for enterprise (see its name :) )

---

note: distinction between returning Err() and throwing
we throw on non-recoverable errors or unexpected errors on our end --- i.e. here we would need to fix the code path that messed up and wrongly called that function. We could btw try to solve it via typing beforehand but it might be overkill, TBD
we return err otherwise